fbpic/particles/spin/inline_functions.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2023, FBPIC contributors
+# Author: Michael J. Quin
+# Scientific supervision: Matteo Tamburini
+# Code optimization: Kristjan Poder
+# License: 3-Clause-BSD-LBNL
+"""
+This file is for use with the Fourier-Bessel Particle-In-Cell code (FB-PIC).
+
+It defines a function to push the 'classical' spin vector, according to the
+Bargmann-Michel-Telegdi equation and other spin-push related functions.
+"""
+import math
+from scipy.constants import c
+
+
+def push_s_BMT(sx_i, sy_i, sz_i, ux_i, uy_i, uz_i, ux_f, uy_f, uz_f,
+               Ex, Ey, Ez, Bx, By, Bz, tauconst, anom):
+    """
+    Push spin of a single particle
+    """
+    # Momentum at a half-timestep forward using Boris assumption
+    ux = (ux_i + ux_f)/2
+    uy = (uy_i + uy_f)/2
+    uz = (uz_i + uz_f)/2
+
+    # Assume 1./gamma at midpoint can be calculated from (average)
+    # momentum at midpoint
+    inv_gamma = 1./math.sqrt( 1. + ux**2 + uy**2 + uz**2 )
+    # Define parameter: 1./(1+gamma)
+    inv_1pgamma = inv_gamma/(1. + inv_gamma)
+    """
+    NOTE! ALGORITHM WRITTEN IN TERMS OF REST FRAME FIELDS.
+    CHECK HOW THE FORM OF BMT EQUATION CHANGES FOR LORENTZ-BOOST.
+    """
+    # Scalar product of momentum and B field
+    uB = (ux*Bx + uy*By + uz*Bz)
+
+    # Define E/B field dependent parameters: tau and upsilon (ups).
+    # Tau vector is analogous to the 'tau' (or 't') vector used in
+    # the Boris-Vay pusher.
+    taux = tauconst*( (anom + inv_gamma)*Bx
+                     - (anom + inv_1pgamma)*(uy*Ez - uz*Ey)*inv_gamma/c
+                     - anom*inv_gamma*inv_1pgamma*uB*ux )
+    tauy = tauconst*( (anom + inv_gamma)*By
+                     - (anom + inv_1pgamma)*(uz*Ex - ux*Ez)*inv_gamma/c
+                     - anom*inv_gamma*inv_1pgamma*uB*uy )
+    tauz = tauconst*( (anom + inv_gamma)*Bz
+                     - (anom + inv_1pgamma)*(ux*Ey - uy*Ex)*inv_gamma/c
+                     - anom*inv_gamma*inv_1pgamma*uB*uz )
+
+    tau2 = taux**2 + tauy**2 + tauz**2
+
+    # # ---------- Boris Method------------------------------------------------
+    # upsx = 2*taux/(1. + tau2)
+    # upsy = 2*tauy/(1. + tau2)
+    # upsz = 2*tauz/(1. + tau2)
+    #
+    # New spin components
+    # sx_f = ( sx_i + sy_i*upsz - sz_i*upsy + (sz_i*taux - sx_i*tauz)*upsz
+    #          - (sx_i*tauy - sy_i*taux)*upsy )
+    # sy_f = ( sy_i + sz_i*upsx - sx_i*upsz + (sx_i*tauy - sy_i*taux)*upsx
+    #          - (sy_i*tauz - sz_i*tauy)*upsz )
+    # sz_f = ( sz_i + sx_i*upsy - sy_i*upsx + (sy_i*tauz - sz_i*tauy)*upsy
+    #          - (sz_i*taux - sx_i*tauz)*upsx )
+    # # -----------------------------------------------------------------------
+
+    # ---------- Tamburini Method ---------------------------------------------
```

Comment:
Does this correspond to the method proposed here:
https://arxiv.org/pdf/2303.16966.pdf
If yes, could you add this reference as a comment?

Reply:
done!